File: club/houseRobber.js
/*
You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed. All houses at this place are arranged in a circle. That means the first house is the neighbor of the last one. Meanwhile, adjacent houses have security system connected and it will automatically contact the police if two adjacent houses were broken into on the same night.
Given a list of non-negative integers representing the amount of money of each house, determine the maximum amount of money you can rob tonight without alerting the police.

Example 1:
Input: [2,3,2]
Output: 3
Explanation: You cannot rob house 1 (money = 2) and then rob house 3 (money = 2), because they are adjacent houses.

Example 2:
Input: [1,2,3,1]
Output: 4
Explanation: Rob house 1 (money = 1) and then rob house 3 (money = 3). Total amount you can rob = 1 + 3 = 4.
*/

// my solution - binary tree
function rob(nums) {
    class BTNode {
        constructor(index, value) {
            this.index = index;
            this.value = value;
            this.skipOne = null;
            this.skipTwo = null;
        }
    };

    class BinaryTree {
        constructor() {
            this.root = null;
        }
        add(index, value) {
            var newNode = new BTNode(index, value);
            if (this.root == null) {
                this.root = newNode;
            }
            else {
                this.addR(this.root, newNode);
            }
            return this;
        }
        addR(node, newNode) {
            if (node.skipOne == null && newNode.index == node.index + 2) {
                node.skipOne = newNode;
            }
            else if (node.skipTwo == null && newNode.index == node.index + 3) {
                node.skipTwo = newNode;
            }
            else if (node.skipOne == null && node.skipTwo == null) {
                return this;
            }
            else {
                this.addR(node.skipOne, newNode);
                newNode = new BTNode(newNode.index, newNode.value);
                this.addR(node.skipTwo, newNode);
            }
            return this;
        }
        sumR(node, num, arr) {
            num += node.value;
            if (node.skipOne != null) {
                this.sumR(node.skipOne, num, arr);
            }
            if (node.skipTwo != null) {
                this.sumR(node.skipTwo, num, arr);
            }
            if (node.skipOne == null && node.skipTwo == null) {
                arr.push(num);
            }
            return arr;
        }
    };

    function robbers(arr) {
        if (arr.length == 0) {
            return 0;
        }
        if (arr.length == 1) {
            return arr[0];
        }
        if (arr.length < 4) {
            var max = arr[0];
            for (var i = 1; i < arr.length; i++) {
                if (arr[i] > max) {
                    max = arr[i];
                }
            }
            return max;
        }
        if (arr.length == 4) {
            var one = arr[0] + arr[2];
            var two = arr[1] + arr[3];
            if (one > two) {
                return one;
            }
            else {
                return two;
            }
        }
        var bt0 = new BinaryTree();
        for (var i = 0; i < arr.length - 1; i++) {
            bt0.add(i, arr[i]);
        }
        var bt1 = new BinaryTree();
        for (var i = 1; i < arr.length; i++) {
            bt1.add(i, arr[i]);
        }
        var bt2 = new BinaryTree();
        for (var i = 2; i < arr.length; i++) {
            bt2.add(i, arr[i]);
        }
        var sum = 0;
        var sumArr = [];
        bt0.sumR(bt0.root, sum, sumArr);
        bt1.sumR(bt1.root, sum, sumArr);
        bt2.sumR(bt2.root, sum, sumArr);
        var max = sumArr[0];
        for (var i = 1; i < sumArr.length; i++) {
            if (sumArr[i] > max) {
                max = sumArr[i];
            }
        }
        return max;

    }
    robbers(nums);
};

// my solution - comparison

// for leetcode
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {

};

File: solved/houseRobber.py
# You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed, the only constraint stopping you from robbing each of them is that adjacent houses have security system connected and it will automatically contact the police if two adjacent houses were broken into on the same night.

# Given a list of non-negative integers representing the amount of money of each house, determine the maximum amount of money you can rob tonight without alerting the police.

# Example 1:
# Input: [1,2,3,1]
# Output: 4
# Explanation: Rob house 1 (money = 1) and then rob house 3 (money = 3).
#              Total amount you can rob = 1 + 3 = 4.

# Example 2:
# Input: [2,7,9,3,1]
# Output: 12
# Explanation: Rob house 1 (money = 2), rob house 3 (money = 9) and rob house 5 (money = 1).
#              Total amount you can rob = 2 + 9 + 1 = 12.

# changes full array into just two variables
class goodSolution:
    def rob(self, nums):
        # if the given array is empty, return zero
        if len(nums) == 0:
            return 0
        # set two variables to store the two highest options
        prev1 = 0
        prev2 = 0
        # iterate through the given array
        for n in nums:
            # store one option in a variable
            temp = prev1
            # update that option by comparing it to the other option plus the current value
            prev1 = max(prev2 + n, temp)
            # update the other option
            prev2 = temp
        # return the highest option
        return prev1

# sill requires an entire array
# aka dynamic programming
class iterativeMemoizationSolution:
    def rob(self, nums):
        # if the given array is empty, there is no total
        if len(nums) == 0:
            return 0
        # utilize an array to track the maximum
        memo = []
        # set the first value as zero, to prevent out-of-range errors
        memo.append(0)
        # set the second value as the first value of the given array, since currently it is the maximum value
        memo.append(nums[0])
        # iterate through the given array
        for i in range(1,len(nums)):
            # compare the current value plus skip two with the skip one value and add the maximum to the maximum array
            memo.append(max(memo[i], memo[i-1] + nums[i]))
        # the final value in the maximum array will be the largest possible amount
        r = memo[len(nums)]
        return r

# has a recursive stack
class recursiveMemoizationSolution:
    def rob(self, nums):
        # utilize an array to remember previous recursive calls
        memo = []
        # fill array with negative numbers
        for i in range(len(nums)+1):
            memo.append(-1)
        # call  recursive function which calculates the maximum between skip one and skip two
        r = self.robr(memo, nums, len(nums)-1)
        return r
    def robr(self, memo, nums, i):
        # base case
        # if i (length of nums array) is less than zero, return zero
        if i < 0:
            return 0
        # base case
        # if the value in the array for i has a zero or positive value, return that amount
        # this is where we prevent calculating the same value multiple times
        if memo[i] >= 0:
            return memo[i]
        # recursively find the maximum between teh current value plus skip two and skip one
        r = max(self.robr(memo, nums, i-2) + nums[i], self.robr(memo, nums, i-1))
        memo[i] = r
        return r

# calculates the same number multiple times
class recursiveSolution:
    def rob(self, nums):
        # call recursive function which calculates the maximum between skip one and skip two
        r = self.robr(nums, len(nums)-1)
        return r
    def robr(self, nums, i):
        # base case
        # if i (length of nums array) is less than zero, return zero
        if i < 0:
            return 0
        # recursively find the maximum between the current value plus skip two and skip one
        r = max(self.robr(nums, i-2) + nums[i], self.robr(nums, i-1))
        return r

nums = [99,98,3,5,6]
s = Solution()
a = s.rob(nums)
print(a)